book/src/generalized-ownership.md

```diff
@@ -0,0 +1,279 @@
+move.
+
+Token program defines ownership of the token to be identified by the public key
+that matches the account's pubkey, and`Token::Transfer`instruction checks
+that the identity of the spending account has a valid signature before
+completing the move.
+
+Both of these programs implement almost the same mechanism.
+
+## Goal for this design
+
+The goal of this design is to make the following possible:
+
+In a single program, it should be possible to change and update ownership of
+arbitrary units of tokens implemented by other programs.
+
+Alice requests X amount of token implemented by program Token1, in
+exchange for Y amount of token implemented by program Token2.
+
+Bob offers X + 1 amount of token implemented by program Token1, in exchange for
+Y - 1 amount of token implemented by Token2.
+
+Dan pairs Bob and Alice, and the result ownership changes:
+
+* Alice owns X of Token1,
+* Bob owns Y - 1 of Token2,
+* Dan owns 1 of Token1 and 1 of Token2,
+
+Another example:
+
+Alice offers X amount of Token1 to the winner of TicTacToe.
+
+Bob offers Y amount of Token2 to the winner of the same game of TicTacToe.
+
+The game charges a commission, the ceiling of the percentage of the tokens.  The
+winner takes ownership of both tokens minus the commission.  The commission is
+transferred to the address designated by the game.
+
+## General idea
+
+For items to have externally countable ownership, they must rely on the external
+program for managing the transfer of units.  These items must have a valid state
+where the number of units owned is 0.  The **zero** count of items is
+the destination for transferring any count of items on the chain.
+
+Execution of these state transitions must allow for a re-entrant and partial
+transactions to make progress.
+
+## Generalized owner identity
+
+Using the account's pubkey identity as the asset owner identity means that all
+the asset transfer APIs must be re-implemented by the program code.
+
+The proposed solution is to use a pointer to an owner identity.
+
+```rust,ignore
+// This state is owned by the Owner program
+enum OwnerState {
+    // the owner identity must provide a valid signature for this pubkey
+    Unique {
+        // The actual owner key.
+        owner: Pubkey, 
+        // The item program instance.
+        item: Pubkey
+    },
+    //...
+}
+```
+
+* OwnerState::Unique.owner - The actual OwnerState’s account’s pubkey is not used to
+distinguish ownership.  Only the OwnerState::Unique.owner pubkey identifies the
+ownership identity.
+
+```rust,ignore
+// This state is owned by the MyUniqueTokenState
+struct MyUniqueTokenState {
+    // This points to an OwnerState structure that identifies the actual owner
+    owner: Pubkey,
+}
+```
+
+* MyUniqueStokenState.owner - address of the OwnerState that identifies the
+current owner.
+
+### OwnerState::Transfer
+
+* account[0] - RW - The OwnerState.
+* account[1] - R - Current owner Pubkey, needs a signature. Doesn’t need an
```

Comment:
A credit-only account needs a signature?

Reply:
Yep, it just needs to show proof of owning the identity.  Kind of interesting, since this account doesn’t need an actual balance

Reply:
Personally, I'd accept tokens from anybody giving tokens out - my signature is not required.